### PR TITLE
Switch true/punycode for symfony/polyfill-intl-idn and update usage.

### PR DESCRIPTION
**Pull request by benbowler**
## Summary

Addresses issue:

- #8094

## Relevant technical choices

Implemented in the IB with fully uppercase `const PUNYCODE_PREFIX = 'xn--';` and set defaults for idn functions: `, IDNA_DEFAULT, INTL_IDNA_VARIANT_UTS46` in order to pass php lint.

## PR Author Checklist

- [x] My code is tested and passes existing unit tests.
- [x] My code has an appropriate set of unit tests which all pass.
- [x] My code is backward-compatible with WordPress 5.2 and PHP 5.6.
- [x] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [x] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [ ] Run the code.
- [ ] Ensure the acceptance criteria are satisfied.
- [ ] Reassess the implementation with the IB.
- [ ] Ensure no unrelated changes are included.
- [ ] Ensure CI checks pass.
- [ ] Check Storybook where applicable.
- [ ] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [ ] Ensure the PR has the correct target branch.
- [ ] Double-check that the PR is okay to be merged.
- [ ] Ensure the corresponding issue has a ZenHub release assigned.
- [ ] Add a changelog message to the issue.

### CODE REVIEW

File: includes/Core/Util/URL.php [hunk 1]
```diff
@@ -147,13 +149,13 @@ public static function permute_site_hosts( $hostname ) {
 			// An ASCII hostname can only be non-IDN or punycode-encoded.
 			if ( $is_ascii ) {
 				// If the hostname is in punycode encoding, add the decoded version to the list of hosts.
-				if ( 0 === strpos( $hostname, Punycode::PREFIX ) || false !== strpos( $hostname, '.' . Punycode::PREFIX ) ) {
-					$host_decoded = $punycode->decode( $hostname );
+				if ( 0 === strpos( $hostname, self::PUNYCODE_PREFIX ) || false !== strpos( $hostname, '.' . self::PUNYCODE_PREFIX ) ) {
+					$host_decoded = idn_to_utf8( $hostname, IDNA_DEFAULT, INTL_IDNA_VARIANT_UTS46 );
```
Comment (techanvil) [inline includes/Core/Util/URL.php:153]: Hmm, looking into it, I think we'd be better off calling `idn_to_utf8()` without specifying the `$flags` and `$variant` args, and instead letting it use the default values.

`$flags` defaults to `IDNA_DEFAULT` anyway, so this is really about the `$variant` argument.

As per the [docs](https://www.php.net/manual/en/function.idn-to-utf8.php), `INTL_IDNA_VARIANT_UTS46` is only available as of ICU 4.6.

However, the `intl` extension, which will be used if it exists, only [depends](https://www.php.net/manual/en/intl.requirements.php) on ICU 4.0+, until PHP 7.4 when it requires 50.1+. With Site Kit supporting PHP 5.6+, SK could well be installed on a system running an ICU version below 4.6.

Wanting to confirm the default `$variant` arg for older versions, I initially looked at the online [legacy docs](https://php-legacy-docs.zend.com/manual/php5/en/function.idn-to-utf8) for v5. However I found it a bit confusing because it refers to version 7 on there too. The [website itself](https://www.php.net/manual/php5.php) says the docs aren't perfect. So I dug through the PHP source, it looks like the default is [`INTL_IDNA_VARIANT_2003` in <=7.3](https://github.com/php/php-src/blob/php-7.3.33/ext/intl/idn/idn.c#L273-L277), and [`INTL_IDNA_VARIANT_UTS46` in 7.4+](https://github.com/php/php-src/blob/php-7.4.0/ext/intl/idn/idn.c#L247-L251).

So, in short it looks like we should be safe to stick with the defaults!

```suggestion
					$host_decoded = idn_to_utf8( $hostname );
```

File: includes/Core/Util/URL.php [hunk 2]
```diff
@@ -147,13 +149,13 @@ public static function permute_site_hosts( $hostname ) {
...
 					array_push( $hosts, $host_decoded, "www.$host_decoded" );
 				}
 			} else {
 				// If it's not ASCII, then add the punycode encoded version.
-				$host_encoded = $punycode->encode( $hostname );
+				$host_encoded = idn_to_ascii( $hostname, IDNA_DEFAULT, INTL_IDNA_VARIANT_UTS46 );
```
Comment (techanvil) [inline includes/Core/Util/URL.php:158]: The same comment applies here, we should stick with the defaults for `$flags` and `$variant`.

```suggestion
				$host_encoded = idn_to_ascii( $hostname );
```

**Review by techanvil — CHANGES_REQUESTED** [review]
Hey @benbowler, this works nicely from what I can see with the tests all passing.

I have left a couple of fairly minor comments, please take a look.

Also, I should have flagged this earlier, but please follow our [branching strategy](https://github.com/google/site-kit-wp/wiki/Engineering#branching-strategy) for future branches and use the issue type label as the first item in the "path". I do actually like having your name embedded in the branch - I will often add my initials to branch myself - but we should generally stick to the defined strategy.

File: composer.json [hunk 3]
```diff
@@ -24,7 +24,7 @@
     "google/apiclient-services": "0.277.*",
     "google/apiclient-services-adsenselinks": "^0.1.0",
     "guzzlehttp/guzzle": "^6.5.8",
-    "true/punycode": "^2.0 <2.1.1"
+		"symfony/polyfill-intl-idn": "<=1.19.0"
```
[the remark this answers is not in the document]
Comment (benbowler) [reply composer.json:27]: @techanvil this throws an error during `composer validate --strict` step of the GitHub actions.

[the remark this answers is not in the document]
Comment (benbowler) [reply composer.json:27]: Speaking to @aaemnnosttv on our weekly catchup, the [composer platform.php config](https://github.com/google/site-kit-wp/blob/ce086ab6966b9be3fafc43f594e4d0eee9ff20ab/composer.json#L44-L46) will ensure that the correct version of the package is installed, therefore I've updated this to `^1`.

File: includes/Core/Util/URL.php (hunk 2 above)
Comment (benbowler) [reply includes/Core/Util/URL.php:158]: Removing these fails `composer lint` with the comment:

> The default value of the idn_to_ascii() $variant parameter has changed from INTL_IDNA_VARIANT_2003 to INTL_IDNA_VARIANT_UTS46 in PHP 7.4. For optimal cross-version compatibility, the $variant parameter should be explicitly set.

I've pushed adding in a `phpcs:ignore` comment so that this function will always use the defaults of the php version it's loaded with.

**Comment by benbowler** [comment]
Thanks @techanvil, I have replied to each comment inline and pushed a new commit to address the comments.

File: composer.json (hunk 3 above)
[the remark this answers is not in the document]
Comment (techanvil) [reply composer.json]: That's great, thanks @benbowler.

File: includes/Core/Util/URL.php [hunk 4]
```diff
@@ -147,13 +149,15 @@ public static function permute_site_hosts( $hostname ) {
 			// An ASCII hostname can only be non-IDN or punycode-encoded.
 			if ( $is_ascii ) {
 				// If the hostname is in punycode encoding, add the decoded version to the list of hosts.
-				if ( 0 === strpos( $hostname, Punycode::PREFIX ) || false !== strpos( $hostname, '.' . Punycode::PREFIX ) ) {
-					$host_decoded = $punycode->decode( $hostname );
+				if ( 0 === strpos( $hostname, self::PUNYCODE_PREFIX ) || false !== strpos( $hostname, '.' . self::PUNYCODE_PREFIX ) ) {
+					// phpcs:ignore PHPCompatibility.ParameterValues.NewIDNVariantDefault.NotSet
```
Comment (techanvil) [inline includes/Core/Util/URL.php:153]: Please can you add a brief comment here explaining why we're disabling the PHPCS rule?

File: includes/Core/Util/URL.php [hunk 5]
```diff
@@ -147,13 +149,15 @@ public static function permute_site_hosts( $hostname ) {
...
+					$host_decoded = idn_to_utf8( $hostname );
 					array_push( $hosts, $host_decoded, "www.$host_decoded" );
 				}
 			} else {
 				// If it's not ASCII, then add the punycode encoded version.
-				$host_encoded = $punycode->encode( $hostname );
+				// phpcs:ignore PHPCompatibility.ParameterValues.NewIDNVariantDefault.NotSet
```
Comment (techanvil) [inline includes/Core/Util/URL.php:159]: Please add a comment here too.

File: includes/Core/Util/URL.php (hunk 2 above)
Comment (techanvil) [reply includes/Core/Util/URL.php]: Top stuff, cheers.

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks @benbowler! I've asked for a couple of inline comments to explain why we're disabling the PHPCS rule, as it's not really evident just looking at the code.

Beyond that this looks good to go!
